getLatestSchemaVersion request failed with error reading entity when refreshing authentication

**Problem description**
During E2E test, I found this problem when testing auth-enabled Pravega.
```
org.apache.flink.util.FlinkRuntimeException: javax.ws.rs.ProcessingException: Error reading entity from input stream.
	at io.pravega.connectors.flink.serialization.DeserializerFromSchemaRegistry.initialize(DeserializerFromSchemaRegistry.java:66)
	at io.pravega.connectors.flink.serialization.DeserializerFromSchemaRegistry.deserialize(DeserializerFromSchemaRegistry.java:109)
	at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEventInternal(EventStreamReaderImpl.java:177)
	at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:125)
	at io.pravega.connectors.flink.FlinkPravegaReader.run(FlinkPravegaReader.java:298)
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:100)
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:63)
	at org.apache.flink.streaming.runtime.tasks.SourceStreamTask$LegacySourceFunctionThread.run(SourceStreamTask.java:213)
Caused by: javax.ws.rs.ProcessingException: Error reading entity from input stream.
	at org.glassfish.jersey.message.internal.InboundMessageContext.readEntity(InboundMessageContext.java:892)
	at org.glassfish.jersey.message.internal.InboundMessageContext.readEntity(InboundMessageContext.java:811)
	at org.glassfish.jersey.client.ClientResponse.readEntity(ClientResponse.java:298)
	at org.glassfish.jersey.client.InboundJaxrsResponse$1.call(InboundJaxrsResponse.java:91)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:205)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:365)
	at org.glassfish.jersey.client.InboundJaxrsResponse.runInScopeIfPossible(InboundJaxrsResponse.java:240)
	at org.glassfish.jersey.client.InboundJaxrsResponse.readEntity(InboundJaxrsResponse.java:88)
	at io.pravega.schemaregistry.client.SchemaRegistryClientImpl.lambda$latestSchemas$8(SchemaRegistryClientImpl.java:231)
	at io.pravega.common.util.Retry$RetryAndThrowBase.run(Retry.java:173)
	at io.pravega.schemaregistry.client.SchemaRegistryClientImpl.withRetry(SchemaRegistryClientImpl.java:482)
	at io.pravega.schemaregistry.client.SchemaRegistryClientImpl.latestSchemas(SchemaRegistryClientImpl.java:229)
	at io.pravega.schemaregistry.client.SchemaRegistryClientImpl.getLatestSchemaVersion(SchemaRegistryClientImpl.java:334)
	at io.pravega.connectors.flink.serialization.DeserializerFromSchemaRegistry.initialize(DeserializerFromSchemaRegistry.java:62)
	... 7 more
Caused by: com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'READ_TABLE': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: UNKNOWN; line: 1, column: 11]
	at com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:1851)
	at com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:717)
	at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._reportInvalidToken(UTF8StreamJsonParser.java:3588)
	at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._handleUnexpectedValue(UTF8StreamJsonParser.java:2683)
	at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._nextTokenNotInObject(UTF8StreamJsonParser.java:865)
	at com.fasterxml.jackson.core.json.UTF8StreamJsonParser.nextToken(UTF8StreamJsonParser.java:757)
	at org.glassfish.jersey.jackson.internal.jackson.jaxrs.base.ProviderBase.readFrom(ProviderBase.java:797)
	at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$TerminalReaderInterceptor.invokeReadFrom(ReaderInterceptorExecutor.java:233)
	at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$TerminalReaderInterceptor.aroundReadFrom(ReaderInterceptorExecutor.java:212)
	at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor.proceed(ReaderInterceptorExecutor.java:132)
	at org.glassfish.jersey.message.internal.MessageBodyFactory.readFrom(MessageBodyFactory.java:1072)
	at org.glassfish.jersey.message.internal.InboundMessageContext.readEntity(InboundMessageContext.java:877)
	... 22 more
```

**Problem location**
`client` and `server`

**Suggestions for an improvement**
Actually there are two problem here:
1. We should first check the status and then read the entity
2. Retries should include authentication exceptions

--- a/client/src/main/java/io/pravega/schemaregistry/client/SchemaRegistryClientImpl.java
+++ b/client/src/main/java/io/pravega/schemaregistry/client/SchemaRegistryClientImpl.java
@@ -228,9 +228,9 @@ public class SchemaRegistryClientImpl implements SchemaRegistryClient {
     private List<SchemaWithVersion> latestSchemas(String groupId, String type) {
         return withRetry(() -> {
             Response response = groupProxy.getSchemas(namespace, groupId, type);
-            SchemaVersionsList objectsList = response.readEntity(SchemaVersionsList.class);
             switch (Response.Status.fromStatusCode(response.getStatus())) {
                 case OK:
+                    SchemaVersionsList objectsList = response.readEntity(SchemaVersionsList.class);
                     return objectsList.getSchemas().stream().map(ModelHelper::decode).collect(Collectors.toList());
                 case NOT_FOUND:
                     throw new ResourceNotFoundException("Group not found.");
@@ -448,9 +448,9 @@ public class SchemaRegistryClientImpl implements SchemaRegistryClient {
     public List<CodecType> getCodecTypes(String groupId) {
         return withRetry(() -> {
             Response response = groupProxy.getCodecTypesList(namespace, groupId);
-            CodecTypes list = response.readEntity(CodecTypes.class);
             switch (Response.Status.fromStatusCode(response.getStatus())) {
                 case OK:
+                    CodecTypes list = response.readEntity(CodecTypes.class);
                     return list.getCodecTypes().stream().map(ModelHelper::decode).collect(Collectors.toList());
                 case NOT_FOUND:
                     throw new ResourceNotFoundException("Group not found.");

--- a/client/src/test/java/io/pravega/schemaregistry/client/TestSchemaRegistryClient.java
+++ b/client/src/test/java/io/pravega/schemaregistry/client/TestSchemaRegistryClient.java
@@ -379,6 +379,9 @@ public class TestSchemaRegistryClient {
         SchemaWithVersion schemaWithVersion1 = client.getLatestSchemaVersion("mygroup", null);
         assertEquals(schemaWithVersion.getSchemaInfo(), schemaWithVersion1.getSchemaInfo());
         assertEquals(schemaWithVersion.getVersionInfo(), schemaWithVersion1.getVersionInfo());
+
+        doThrow(new RuntimeException()).when(response).readEntity(
+                SchemaVersionsList.class);
         // NotFound Exception
         doReturn(Response.Status.NOT_FOUND.getStatusCode()).when(response).getStatus();
         AssertExtensions.assertThrows("An exception should have been thrown",
@@ -388,6 +391,8 @@ public class TestSchemaRegistryClient {
         AssertExtensions.assertThrows("An exception should have been thrown",
                 () -> client.getLatestSchemaVersion("mygroup", null), e -> e instanceof InternalServerError);
 
+        doReturn(schemaWithVersions).when(response).readEntity(
+                SchemaVersionsList.class);
         doReturn(Response.Status.OK.getStatusCode()).when(response).getStatus();
         serializationFormat = SerializationFormat.custom("custom");
         versionInfo = new VersionInfo("schema2", serializationFormat.getFullTypeName(), 5, 5);

--- a/server/src/main/java/io/pravega/schemaregistry/storage/client/TableStore.java
+++ b/server/src/main/java/io/pravega/schemaregistry/storage/client/TableStore.java
@@ -94,6 +94,10 @@ public class TableStore extends AbstractService {
     }
     
     TableStore(WireCommandClient wireCommandClient, ScheduledExecutorService executor) {
+        this(wireCommandClient, executor, NUM_OF_RETRIES);    
+    }
+    
+    TableStore(WireCommandClient wireCommandClient, ScheduledExecutorService executor, int retryCount) {
         this.wireCommandClient = wireCommandClient;
         this.hostStore = wireCommandClient.getHostStore();
         this.executor = executor;
@@ -101,7 +105,7 @@ public class TableStore extends AbstractService {
             String[] splits = x.split("/");
             return hostStore.getController().getOrRefreshDelegationTokenFor(splits[0], splits[1], AccessOperation.READ_WRITE).join();
         };
-        numOfRetries = NUM_OF_RETRIES;
+        numOfRetries = retryCount;
         this.cache = CacheBuilder.newBuilder()
                                  .maximumSize(Config.TABLE_SEGMENT_CACHE_SIZE)
                                  .build();
@@ -380,8 +384,11 @@ public class TableStore extends AbstractService {
     private <T> CompletableFuture<T> withRetries(Supplier<CompletableFuture<T>> futureSupplier, Supplier<String> errorMessage,
                                                  String tableName) {
         return Retry.withExpBackoff(RETRY_INIT_DELAY, RETRY_MULTIPLIER, numOfRetries, RETRY_MAX_DELAY)
-                    .retryWhen(e -> Exceptions.unwrap(e) instanceof StoreExceptions.StoreConnectionException)
-                    .runAsync(exceptionalCallback(futureSupplier, errorMessage, tableName), executor);
+                    .retryWhen(e -> {
+                        Throwable unwrap = Exceptions.unwrap(e);
+                        return unwrap instanceof StoreExceptions.StoreConnectionException || 
+                                unwrap instanceof StoreExceptions.TokenException;
+                    }).runAsync(exceptionalCallback(futureSupplier, errorMessage, tableName), executor);
     }
     
     @SneakyThrows(ExecutionException.class)

--- a/server/src/test/java/io/pravega/schemaregistry/storage/client/TablesStoreTest.java
+++ b/server/src/test/java/io/pravega/schemaregistry/storage/client/TablesStoreTest.java
@@ -14,6 +14,7 @@ import io.netty.buffer.ByteBuf;
 import io.netty.buffer.Unpooled;
 import io.pravega.common.Exceptions;
 import io.pravega.common.concurrent.ExecutorServiceHelpers;
+import io.pravega.common.util.RetriesExhaustedException;
 import io.pravega.schemaregistry.ResultPage;
 import io.pravega.schemaregistry.storage.StoreExceptions;
 import io.pravega.test.common.AssertExtensions;
@@ -31,6 +32,10 @@ import java.util.concurrent.ScheduledExecutorService;
 import java.util.stream.Collectors;
 
 import static org.junit.Assert.*;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyBoolean;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
 
 public class TablesStoreTest {
     private static final String SCHEMAREGISTRY_TABLE = "schemaregistry/table";
@@ -171,5 +176,31 @@ public class TablesStoreTest {
         assertEquals(keys.get(1), new String(values.get(1).getRecord()));
         assertSame(values.get(2).getVersion().toLong(), Version.NON_EXISTENT.toLong());
         assertEquals(keys.get(3), new String(values.get(3).getRecord()));
+    }
+
+    @Test
+    public void retryTest() {
+        // connection exception
+        WireCommandClient wireCommandClient = WireCommandMock.getFailingMock(() -> 
+                StoreExceptions.create(StoreExceptions.Type.CONNECTION_ERROR, "my exception"));
+        TableStore tableStore = new TableStore(wireCommandClient, executor, 2);
+        AssertExtensions.assertFutureThrows("retries did not exhaust", 
+                tableStore.createTable("t/t"), e -> Exceptions.unwrap(e) instanceof RetriesExhaustedException);
+        verify(wireCommandClient, times(2)).createTableSegment(any(), any());
+
+        // auth error
+        wireCommandClient = WireCommandMock.getFailingMock(() -> 
+                StoreExceptions.create(StoreExceptions.Type.AUTH_ERROR, "my exception"));
+        tableStore = new TableStore(wireCommandClient, executor, 2);
+        AssertExtensions.assertFutureThrows("retries did not exhaust",
+                tableStore.deleteTable("t/t", true), e -> Exceptions.unwrap(e) instanceof RetriesExhaustedException);
+        verify(wireCommandClient, times(2)).deleteTableSegment(any(), anyBoolean(), any());
+
+        // general error, no retries should be performed
+        wireCommandClient = WireCommandMock.getFailingMock(RuntimeException::new);
+        tableStore = new TableStore(wireCommandClient, executor, 2);
+        AssertExtensions.assertFutureThrows("no retries should be performed",
+                tableStore.removeEntry("t/t", new byte[0]), e -> Exceptions.unwrap(e) instanceof RuntimeException);
+        verify(wireCommandClient, times(1)).removeTableKeys(any(), any(), any());
     }
 }

--- a/server/src/test/java/io/pravega/schemaregistry/storage/client/WireCommandMock.java
+++ b/server/src/test/java/io/pravega/schemaregistry/storage/client/WireCommandMock.java
@@ -18,6 +18,7 @@ import io.pravega.client.tables.impl.IteratorStateImpl;
 import io.pravega.client.tables.impl.TableSegmentEntry;
 import io.pravega.client.tables.impl.TableSegmentKey;
 import io.pravega.client.tables.impl.TableSegmentKeyVersion;
+import io.pravega.common.concurrent.Futures;
 import io.pravega.common.util.BitConverter;
 import io.pravega.controller.stream.api.grpc.v1.Controller;
 import io.pravega.schemaregistry.storage.StoreExceptions;
@@ -34,6 +35,7 @@ import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
 import java.util.stream.Collectors;
 
 import static org.mockito.ArgumentMatchers.any;
@@ -288,6 +290,55 @@ public class WireCommandMock {
                 }
             }, executor);
         }).when(helper).readTableEntries(anyString(), anyInt(), any(), anyString());
+        // endregion
+        
+        return helper;
+    }
+    
+    public static WireCommandClient getFailingMock(Supplier<? extends Exception> exceptionSupplier) {
+        WireCommandClient helper = spy(new WireCommandClient(mock(ConnectionPool.class), mock(HostStore.class)));
+
+        HostStore hoststore = mock(HostStore.class);
+        
+        doReturn(CompletableFuture.completedFuture(Controller.NodeUri.newBuilder().setEndpoint("localhost").setPort(SERVICE_PORT).build()))
+                .when(hoststore).getHostForTableSegment(anyString());
+
+        ControllerImpl controller = mock(ControllerImpl.class);
+        doReturn(CompletableFuture.completedFuture("")).when(controller).getOrRefreshDelegationTokenFor(anyString(), anyString(), any());
+        doReturn(CompletableFuture.completedFuture(true)).when(controller).createScope(anyString());
+
+        doReturn(controller).when(hoststore).getController();
+        doReturn(hoststore).when(helper).getHostStore();
+        
+        doReturn(CompletableFuture.completedFuture(Controller.NodeUri.newBuilder().setEndpoint("localhost").setPort(SERVICE_PORT).build()))
+                .when(helper).getTableUri(anyString());
+
+        // region create table
+        doAnswer(x -> Futures.failedFuture(exceptionSupplier.get())).when(helper).createTableSegment(anyString(), anyString());
+        // endregion
+        
+        // region delete table
+        doAnswer(x -> Futures.failedFuture(exceptionSupplier.get())).when(helper).deleteTableSegment(anyString(), anyBoolean(), anyString());
+        // endregion
+        
+        // region update keys
+        doAnswer(x -> Futures.failedFuture(exceptionSupplier.get())).when(helper).updateTableEntries(anyString(), any(), anyString());
+        // endregion
+    
+        // region remove keys    
+        doAnswer(x -> Futures.failedFuture(exceptionSupplier.get())).when(helper).removeTableKeys(anyString(), any(), anyString());
+        // endregion
+
+        // region read keys    
+        doAnswer(x -> Futures.failedFuture(exceptionSupplier.get())).when(helper).readTable(anyString(), any(), anyString());
+        // endregion
+        
+        // region readTableKeys
+        doAnswer(x -> Futures.failedFuture(exceptionSupplier.get())).when(helper).readTableKeys(anyString(), anyInt(), any(), anyString());
+        // endregion        
+        
+        // region readTableEntries
+        doAnswer(x -> Futures.failedFuture(exceptionSupplier.get())).when(helper).readTableEntries(anyString(), anyInt(), any(), anyString());
         // endregion
         
         return helper;

